### PR TITLE
Update 20170704165711_add_user_bio_and_token.rb (fix for #1510)

### PR DESCRIPTION
**Pull request by icarito**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [ ] all tests pass -- `rake test:all`
* [ ] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [ ] pull request is descriptively named with #number reference back to original issue

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by icarito** [comment]
After this the migration fails differently, for me:
```
icarito@tycho:/srv/plots_testing/plots$ RAILS_ENV=production docker-compose run web rake db:migrate
Starting plots_db_1 ... done
==  AddUserBioAndToken: migrating =============================================
-- add_column(:rusers, :bio, :text, {:limit=>2147483647})
   -> 20.0193s
-- add_column(:rusers, :token, :string)
   -> 26.2858s
-- add_column(:rusers, :status, :integer, {:default=>0})
   -> 26.1703s
-- remove_column(:rusers, :location_privacy)
   -> 27.6474s
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '=': SELECT  `rusers`.* FROM `rusers`  WHERE `rusers`.`username` = 'トリーバーチ クラッチ' LIMIT 1/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
ActiveRecord::StatementInvalid: Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '=': SELECT  `rusers`.* FROM `rusers`  WHERE `rusers`.`username` = 'トリーバーチ クラッチ' LIMIT 1
/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='
/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

**Comment by icarito** [comment]
Hi,
I've worked around the other issues blocking the migration. I'm just not sure if the changes are acceptable, please review (force collation and ignore rusers without status).

**Comment by icarito** [comment]
I've made the migration in staging manually, btw, that's how I tested it.

**Comment by icarito** [comment]
Ouch! This appeared to work, but seems to break everything else, according to Travis.
That's all I have for now. Good night!

File: app/models/drupal_users.rb [hunk 1]
```diff
@@ -29,7 +29,7 @@ def internalShouldIndexSolr
   end
 
   def user
-    User.find_by_username name
+    User.where("username collate latin1_swedish_ci = '" + name + "'")
```
Comment (jywarren) [inline app/models/drupal_users.rb]: Whoa, this is intense. do we need to do this elsewhere? Do we need to limit character set allowed?

**Comment by jywarren** [comment]
.where returns an array like collection, so I added a .first to the end. Let's see!

**Comment by jywarren** [comment]
Got a failure as the collate seemed incompatible w utf8, so trying this...

**Comment by jywarren** [comment]
This looks good! I'm going to merge so we can try pushing it to staging.

**Comment by jywarren** [comment]
Hmm, something broke! http://jenkins.laboratoriopublico.org/job/Plots-Staging/145/console

**Comment by icarito** [comment]
It looks like a permissions issue, I'll reset permissions and rebuild.

**Comment by icarito** [comment]
Now we should see it! Keep in mind migration was done yesterday, perhaps it's necessary to roll it back and redo it?

**Comment by jywarren** [comment]
possibly... this error snuck by Travis, but in general we shouldn't have to
modify a migration on a persistent instance... always adding a new
migration. But this time, would you mind rolling back to previous version
and rerunning?

**Comment by jywarren** [comment]
hmm, but I see an error here now, unfortunately -- any idea why? http://staging.laboratoriopublico.org/dashboard

**Comment by icarito** [comment]
I see this in the logs:
```
^[[36mweb_1  |^[[0m Started GET "/tag/question:biology" for 137.74.203.194 at 2017-07-08 18:25:02 +0000
^[[36mweb_1  |^[[0m Processing by TagController#show as */*
^[[36mweb_1  |^[[0m   Parameters: {"id"=>"question:biology"}
^[[36mweb_1  |^[[0m   Rendered questions/_questions.html.erb (2.6ms)
^[[36mweb_1  |^[[0m   Rendered tag/show.html.erb within layouts/application (3.9ms)
^[[36mweb_1  |^[[0m Completed 500 Internal Server Error in 59.4ms
^[[36mweb_1  |^[[0m
^[[36mweb_1  |^[[0m ActionView::Template::Error (Mysql2::Error: Unknown column 'term_data.parent' in 'where clause': SELECT  DISTINCT `node`.nid FROM `node` LEFT OUTER JOIN `node_revisions` ON `node_revisions`.`nid` = `node`.`nid` LEFT OUTER JOIN `community_tags` ON `community_tags`.`nid` = `node`.`nid` LEFT OUTER JOIN `term_data` ON `term_data`.`tid` = `community_tags`.`tid` WHERE `node`.`status` = 1 AND `node`.`type` = 'note' AND (term_data.name = 'question:biology' OR term_data.parent = 'question:biology') AND (node.nid IN (10469,10537,10574,10984,10993,10996,11027,11135,11140,11185,11226,11270,11301,11311,11312,11327,11378,11406,11435,11480,11516,11524,11526,11545,11610,11633,11635,11686,11710,11713,11717,11718,11730,11733,11836,11845,11946,11953,11962,11984,12006,12043,12068,12127,12129,12135,12155,12156,12215,12231,12251,12304,12311,12339,12341,12365,12436,12450,12482,12548,12549,12607,12635,12657,12700,12701,12703,12767,12820,12843,12882,12965,12978,12985,13000,13025,13027,13042,13055,13056,13061,13081,13082,13115,13172,13190,13233,13234,13247,13252,13256,13286,13312,13356,13358,13359,13360,13362,13363,13371,13385,13388,13411,13418,13424,13425,13427,13446,13447,13451,13452,13453,13454,13456,13461,13467,13482,13498,13499,13500,13505,13506,13509,13517,13527,13538,13540,13544,13548,13549,13550,13552,13553,13556,13557,13572,13575,13635,13636,13638)) ORDER BY node_revisions.timestamp DESC LIMIT 30 OFFSET 0):
```
Unknown column...
It could be that I messed up the staging database, so I'll copy a fresh one from production later today or tomorrow. I notice not every migration can be rolled back and replayed.

**Comment by jywarren** [comment]
Hmm that's odd. I think one thing is that if a migration fails, it
unfortunately doesn't roll back automatically. I wish it worked in a
transaction!

On Jul 8, 2017 2:33 PM, "Sebastian Silva" <notifications@github.com> wrote:

I see this in the logs:

^[[36mweb_1  |^[[0m Started GET "/tag/question:biology" for
137.74.203.194 at 2017-07-08 18:25:02 +0000
^[[36mweb_1  |^[[0m Processing by TagController#show as */*
^[[36mweb_1  |^[[0m   Parameters: {"id"=>"question:biology"}
^[[36mweb_1  |^[[0m   Rendered questions/_questions.html.erb (2.6ms)
^[[36mweb_1  |^[[0m   Rendered tag/show.html.erb within
layouts/application (3.9ms)
^[[36mweb_1  |^[[0m Completed 500 Internal Server Error in 59.4ms
^[[36mweb_1  |^[[0m
^[[36mweb_1  |^[[0m ActionView::Template::Error (Mysql2::Error:
Unknown column 'term_data.parent' in 'where clause': SELECT  DISTINCT
`node`.nid FROM `node` LEFT OUTER JOIN `node_revisions` ON
`node_revisions`.`nid` = `node`.`nid` LEFT OUTER JOIN `community_tags`
ON `community_tags`.`nid` = `node`.`nid` LEFT OUTER JOIN `term_data`
ON `term_data`.`tid` = `community_tags`.`tid` WHERE `node`.`status` =
1 AND `node`.`type` = 'note' AND (term_data.name = 'question:biology'
OR term_data.parent = 'question:biology') AND (node.nid IN
(10469,10537,10574,10984,10993,10996,11027,11135,11140,11185,11226,11270,11301,11311,11312,11327,11378,11406,11435,11480,11516,11524,11526,11545,11610,11633,11635,11686,11710,11713,11717,11718,11730,11733,11836,11845,11946,11953,11962,11984,12006,12043,12068,12127,12129,12135,12155,12156,12215,12231,12251,12304,12311,12339,12341,12365,12436,12450,12482,12548,12549,12607,12635,12657,12700,12701,12703,12767,12820,12843,12882,12965,12978,12985,13000,13025,13027,13042,13055,13056,13061,13081,13082,13115,13172,13190,13233,13234,13247,13252,13256,13286,13312,13356,13358,13359,13360,13362,13363,13371,13385,13388,13411,13418,13424,13425,13427,13446,13447,13451,13452,13453,13454,13456,13461,13467,13482,13498,13499,13500,13505,13506,13509,13517,13527,13538,13540,13544,13548,13549,13550,13552,13553,13556,13557,13572,13575,13635,13636,13638))
ORDER BY node_revisions.timestamp DESC LIMIT 30 OFFSET 0):

Unknown column...
It could be that I messed up the staging database, so I'll copy a fresh one
from production later today or tomorrow. I notice not every migration can
be rolled back and replayed.

—
You are receiving this because you modified the open/close state.

Reply to this email directly, view it on GitHub
<https://github.com/publiclab/plots2/pull/1511#issuecomment-313873506>, or mute
the thread

**Comment by icarito** [comment]
What I meant when I said I could've messed up the schema is that at one point I tried doing `rake db:rollback STEP=20170704165711` thinking I could rollback a specific step, instead it started rolling back up until I hit Ctrl-C. When I attempted to replay migrations, I hit `[db/migrate/20161122154603_remove_friendly_id_indices.rb](https://github.com/publiclab/plots2/blob/stable/db/migrate/20161122154603_remove_friendly_id_indices.rb)` which doesn't have a proper rollback method, so it was attempting to remove an index which didn't exist. Not knowing I could've just commented it out, I went ahead and created the indexes so that the migration could remove them and continue.
I haven't gone thru every migration to determine if there were any other side effects.
I'm going to `mysqldump` and copy a fresh database from production into staging to be safe.

**Comment by icarito** [comment]
Hi,
I did the migration and got this:
```
icarito@tycho:/srv/plots_testing/plots$ RAILS_ENV=production docker-compose run web rake db:migrate
Starting plots_db_1 ... done
==  AddUserBioAndToken: migrating =============================================
-- add_column(:rusers, :bio, :text, {:limit=>2147483647})
   -> 23.2842s
-- add_column(:rusers, :token, :string)
   -> 30.4798s
-- add_column(:rusers, :status, :integer, {:default=>0})
   -> 31.6240s
-- remove_column(:rusers, :location_privacy)
   -> 31.7525s                                                                                                                                rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '=': SELECT  `rusers`.* FROM `rusers`  WHERE `rusers`.`username` = 'トリーバーチ クラッチ' LIMIT 1/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
ActiveRecord::StatementInvalid: Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '=': SELECT  `rusers`.* FROM `rusers`  WHERE `rusers`.`username` = 'トリーバーチ クラッチ' LIMIT 1
/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
Mysql2::Error: Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='
/app/app/models/drupal_users.rb:32:in `user'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:10:in `block in up'
/app/db/migrate/20170704165711_add_user_bio_and_token.rb:9:in `up'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```
I notice you removed the collation bits from `app/models/drupal_users.rb`, but that was the point of the edit...

**Comment by jywarren** [comment]
I had gotten an error that said it wasn't allowed in MySQL, I think...
Should be in error log of a previous build. But I'm not sure the solution
then. I'm a little surprised there are non Latin characters; aren't they
forbidden by our DrupalUsers validations? Or is this a real user and
they're using that username?

**Comment by icarito** [comment]
I'm manually undoing the last migration step, so that next push will have a clean schema.

**Comment by icarito** [comment]
The db is awaiting another try, the last time I was able to do it with the `collate` parameter.

**Comment by icarito** [comment]
@jywarren let me know if I can help. The database is up to date with a copy from yesterday (pending this last migration).
If you find migration breaks, I went into the staging db with the command:
`tycho:/srv/plots_testing/plots$ RAILS_ENV=production docker-compose exec db mysql -u plots -p`
(password is "`plots`", as set in docker-compose.yml file)

**Comment by jywarren** [comment]
OK, opened a possible solution in #1514

**Pull request merged**